### PR TITLE
[FLINK-1901] [core] Create sample operator for Dataset.

### PR DESCRIPTION
**Pull request by ChengXiangLi**
This PR includes:
1. 4 random sampler implementation for different sample strategies.
2. sample operator for DataSet Java API.
3. random sampler unit test.
4. sample operator Java API integration test.

### CODE REVIEW

**Comment by ChengXiangLi** [comment]
Previously, i plan to leave the sample scala API to an separate PR as i not very familiar with scala, but the failed test shows that Flink has a test to make sure scala and java has the same API, i would try to add scala API and integration test later.

**Comment by ChengXiangLi** [comment]
scala API and it's integration test has been merged into latest commit. @tillrohrmann do you have time to review this PR?

**Comment by tillrohrmann** [comment]
Thanks for your contribution @ChengXiangLi. I'll try to review your PR tomorrow morning.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+						T element = input.next();
+						if (random.nextDouble() <= fraction) {
+							current = element;
+							return true;
+						}
+					}
+					current = null;
+					return false;
+				}
+				return false;
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java]: I think, if I'm not mistaken, that `hasNext` has to be idempotent. Thus it should return `true` if `current != null`.

Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java]: Could we put the last return statement in an `else` branch?

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ */
+package org.apache.flink.api.common.operators.util;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.math3.distribution.PoissonDistribution;
+
+import java.util.Iterator;
+
+/**
+ * A sampler implementation based on Poisson Distribution. While sample elements with replacement,
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java]: "While sampling elements"

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+ * the picked number of each element follow poisson distribution, so we could use poisson distribution
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java]: Typo: "element follows a given poisson distribution"

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/RandomSampler.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.common.operators.util;
+
+import java.util.Iterator;
+
+public abstract class RandomSampler<T> {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/RandomSampler.java]: JavaDocs are missing

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/RandomSampler.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	/**
+	 * Randomly sample the elements from input, and return the result iterator.
+	 *
+	 * @param input source data
+	 * @return the sample result.
+	 */
+	public abstract Iterator<T> sample(Iterator<T> input);
+}
+
+abstract class SampledIterator<T> implements Iterator<T> {
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/RandomSampler.java]: JavaDocs are missing

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * A simple in memory implementation of Reservoir Sampling with replacement, and with only one pass through
+ * the input iteration whose size is unpredictable.
+ * This implementation refers to the algorithm described in "Reservoir-based Random Sampling with Replacement
+ * from Data Stream".
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java]: Maybe put a link to the paper in here.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	 * @param seed       random number generator seed
+	 */
+	public ReservoirSamplerWithReplacement(int numSamples, long seed) {
+		this(numSamples, new Random(seed));
+	}
+	
+	/**
+	 * Create a reservoir sampler with fixed sample size and random number generator.
+	 *
+	 * @param numSamples number of samples to retain in reservoir, must be non-negative.
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java]: "@param" is missing

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithoutReplacement.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Random;
+
+/**
+ * A simple in memory implementation of Reservoir Sampling without replacement, and with only one pass through
+ * the input iteration whose size is unpredictable.
+ * This implementation refers to the Algorithm R described in "Random Sampling with a Reservoir" Vitter, 1985.
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithoutReplacement.java]: Maybe add again a link to the paper if available.

File: pom.xml [hunk 9]
```diff
@@ -224,6 +224,12 @@ under the License.
 				<version>3.2.1</version>
 			</dependency>
 
+			<dependency>
+				<groupId>org.apache.commons</groupId>
+				<artifactId>commons-math3</artifactId>
+				<version>3.5</version>
+			</dependency>
```
Comment (tillrohrmann) [inline pom.xml]: For that we have to add an entry in the `flink-dist/NOTICE` and `flink-dist/LICENSE` files. But I can do that when merging the PR.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @ChengXiangLi. The code is really well tested and well structured. Great work :-)

I had only some minor comments. There is however one thing I'm not so sure about. With the current implementation, all parallel tasks of the sampling operator will get the same random generator/seed value. Thus, every node will generate the same sequence of random numbers. I think this can have a negative influence on the sampling. What we could do is to use `RichMapPartitionFunction` instead of the `MapPartitionFunction`. With the rich function, we either have access to the subtask index, given by `getRuntimeContext().getIndexOfThisSubtask()`,  which we could use to modify the initial seed or we generate the random number generator in the `open` method (this method is executed on the TaskManager). Assuming that the clocks are not completely synchronized and that the individual tasks will be instantiated not at the same time, this could give us less correlated random number sequences. What do you think?

**Comment by ChengXiangLi** [comment]
Thanks for the review, Till. For your last suggestion, I'm aware of this issue before, but i didn't know the exist of RichMapPartitionFunction, thanks for reminding it, i would update the code soon.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java (hunk 7 above)
Comment (ChengXiangLi) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/ReservoirSamplerWithReplacement.java]: Fixed.

File: pom.xml (hunk 9 above)
Comment (ChengXiangLi) [inline pom.xml]: I didn't found flink-dist/NOTICE and flink-dist/LICENSE files, there are NOTICE and LICENSE in flink root directory, but i didn't found the right place to add something related to commons-math3, so i would just leave this to you, Till.

**Comment by ChengXiangLi** [comment]
Hi, @tillrohrmann , current implementation of sample with fixed size would generate fixed size sample for each partition randomly instead of the whole dataset, user may expect the later one actually most of the time. I'm research on how to sample fixed size elements randomly from distributed data stream, i think we can pause this PR review until i merge the previous fix.

File: pom.xml (hunk 9 above)
Comment (tillrohrmann) [inline pom.xml]: Sorry, they are actually located in `flink-dist/src/main/flink-bin`. But I'll update them when I merge the PR.

**Comment by tillrohrmann** [comment]
The current state with the `RichMapPartitionFunctions` looks good to me :+1: 

You're right that user usually want to fix the size for the whole sample. An easy solution could be to assign each item an index, see `DataSetUtils.zipWithIndex`. Then we can compute the maximum index (which is effectively counting the data set elements). This gives us the range from which have to sample. By generating a parallel sequence of the size of our sample size with `env.generateSequence(maxIndex)`, we could then sample from `[0, maxIndex]`. At last we would have to join this data set with the original data set which has the indices assigned. There are probably more efficient algorithms out there than this one.

Just ping me when you've found a solution for the problem. Looking forward reviewing it :-)

**Comment by sachingoel0101** [comment]
I have worked on this problem before. The idea is to divide the data into blocks and find the probability of selection of an element from a block.
Thus, suppose there are blocks B_1, B_2, ..., B_N with probabilities P_1, P_2, ..., P_N, then you sample k points by first sampling from the distribution {P_1, P_2, ..., P_N} and find the number of elements you require from each block. After that you select the required number of points from each block and take a union.
It is pretty easy to implement in a shared memory system but with a distributed system, it is hard. I tried the following approach some time before, although didn't quite finish working on it:
blockedData = Data -> (block_id, data)

blockNumbers = (block_id, data) -> (block_id, count)

(1...k) -> (list of block ids we'll be sampling from)
After this, I tried broadcasting the list and selecting the required number of elements from each block, which can be done quite easily. But what if k is very large?

**Comment by ChengXiangLi** [comment]
Thanks for the input, @tillrohrmann and @sachingoel0101 . I would like to implement the fixed size sampling with only one pass through source dataset, since while user try to sample a dataset, the dataset should be quite large in most cases, pass through the dataset multi times would add much more effort. In my solution, the basic idea of fixed size sample in distributed stream is that: generate a random number for each input elements as its weight, select top K elements with max weight, as the weights are generated randomly, so the selected top K elements are selected randomly. You can see more detail information in the code and javadoc.

**Comment by thvasilo** [comment]
Hello @ChengXiangLi, perhaps looking at [this](http://machinelearning.wustl.edu/mlpapers/papers/icml2013_meng13a) paper may help with deciding which sampling algorithm to use when deciding for the exact sample size algorithm.

It provides an implementation specifically designed for a MapReduce environment.

**Comment by ChengXiangLi** [comment]
Thanks, @thvasilo , that paper introduced an random sample algorithm which is an extend algorithm of the one i described before, it has two threshold the filter the element before sort, if element weight is bigger than "up threshold", it would be included in final top K elements with very high possibility, if element weight is smaller than "down threshold", it would not be included in final top K elements with very high possibility. With accepted possibility, we can filter the element with weigh larger than "up threshold" or smaller than "down threshold", only sort the elements with weight between the thresholds.
 
This is a very good algorithm, i would add it on my notebook for further improvement, but i don't want to implement it right way. This PR is large enough to me, so i would like to leave all the algorithms optimization in future, and just keep the basic implementations of sample algorithms here, make sure they are simple, easy to understand, work correctly, and they can be used as the performance base line for the further improvement.

**Comment by thvasilo** [comment]
Agreed, we can take a look at the optimized algorithm later.

**Comment by ChengXiangLi** [comment]
Hi, @tillrohrmann , it's ready for review now.

File: flink-core/src/test/java/org/apache/flink/api/common/operators/util/RandomSamplerTest.java [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+	public void testReservoirSamplerWithMultiSourcePartitions2() {
+		initSourcePartition();
+
+		verifyReservoirSamplerWithReplacement(100, true);
+		verifyReservoirSamplerWithReplacement(500, true);
+		verifyReservoirSamplerWithReplacement(1000, true);
+		verifyReservoirSamplerWithReplacement(5000, true);
+	}
+	
+	private void verifySamplerFixedSampleSize(int numSample, boolean withReplacement) {
```
Comment (thvasilo) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/RandomSamplerTest.java]: Having some comments for some of the ``verify[..]`` functions explaining what we verify at each would help with code understanding.

Comment (ChengXiangLi) [inline flink-core/src/test/java/org/apache/flink/api/common/operators/util/RandomSamplerTest.java]: Thanks, @thvasilo , I've add more comments in the latest commit.

File: flink-scala/src/main/scala/org/apache/flink/api/scala/DataSet.scala [hunk 11]
```diff
@@ -1182,6 +1184,60 @@ class DataSet[T: ClassTag](set: JavaDataSet[T]) {
...
+      seed: Long = Utils.RNG.nextLong()): DataSet[T] = {
+
+    wrap(new MapPartitionOperator[T, T](javaSet,
+      getType(),
+      new SampleWithFraction(withReplacement, fraction, seed),
+      getCallLocationName()))
+  }
+
+  /**
+   * Generate a sample of DataSet by the probability fraction of each element.
```
Comment (thvasilo) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/DataSet.scala]: Javadoc is from the fraction function.

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+package org.apache.flink.api.common.operators.util;
+
+import com.google.common.base.Preconditions;
+
+import java.util.Iterator;
+import java.util.Random;
+
+/**
+ * A sampler implementation built upon Bernoulli Trail. For sample with fraction and without replacement,
+ * each element sample choice is just a bernoulli trail.
```
Comment (thvasilo) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java]: Do you mean Bernouli _trial_ here?

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * @param <T> The type of sample.
+ */
+public class BernoulliSampler<T> extends RandomSampler<T> {
+	
+	private final double fraction;
+	private final Random random;
+	
+	/**
+	 * Create a bernoulli sampler sample fraction and default random number generator.
```
Comment (thvasilo) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java]: *B*ernouli should be capitalized for all mentions

File: flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+  @throws(classOf[Exception])
+  private def verifySamplerWithFractionWithReplacement(fraction: Double, seed: Long) {
+    verifySamplerWithFraction(true, fraction, seed)
+  }
+
+  @throws(classOf[Exception])
+  private def verifySamplerWithFraction(withReplacement: Boolean, fraction: Double, seed: Long) {
+    val ds = getSourceDataSet()
+    val sampled = ds.sample(withReplacement, fraction, seed)
+    result = sampled.collect.asJava
```
Comment (thvasilo) [inline flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala]: Is this result checked for validity somewhere?

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+					}
+					current = null;
+					return false;
+				} else {
+					return true;
+				}
+			}
+			
+			@Override
+			public T next() {
```
Comment (thvasilo) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/BernoulliSampler.java]: It feels a bit counterintuitive that the next element is prepared in the `hasNext()` function. Doesn't this mean that `hasNext()` **needs** to be called every time before we call `next()`?

Can we protect against that case where we would get a `null` element back that way?

File: flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java [hunk 16]
```diff
@@ -451,6 +454,53 @@ protected static File asFile(String path) {
 			assertEquals(extectedStrings[i], resultStrings[i]);
 		}
 	}
+	
+	// --------------------------------------------------------------------------------------------
+	// Comparison methods for tests using sample
+	// --------------------------------------------------------------------------------------------
+	
+	public static <T> void containsResultAsTuples(List<T> result, String expected) {
```
Comment (thvasilo) [inline flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java]: Is this used anywhere?

File: flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java [hunk 17]
```diff
@@ -451,6 +454,53 @@ protected static File asFile(String path) {
...
+		isExpectedContainsResult(result, expected, true);
+	}
+	
+	public static <T> void containsResultAsText(List<T> result, String expected) {
+		isExpectedContainsResult(result, expected, false);
+	}
+	
+	private static <T> void isExpectedContainsResult(List<T> result, String expected, boolean asTuple) {
```
Comment (thvasilo) [inline flink-test-utils/src/main/java/org/apache/flink/test/util/TestBaseUtils.java]: Can we get comments explaining the functionality of this and `containsResultAsText`?

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 18]
```diff
@@ -1057,7 +1061,68 @@ public long count() throws Exception {
...
+		return mapPartition(new SampleWithFraction<T>(withReplacement, fraction, seed));
+	}
+	
+	/**
+	 * Generate a sample of DataSet which contains fixed size elements.
+	 *
+	 * @param withReplacement Whether element can be selected more than once.
+	 * @param numSample       The expected sample size.
+	 * @return The sampled DataSet
+	 */
```
Comment (thvasilo) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: Maybe we want to include a note that this kind of sampling currently takes 2 passes over the data, and recommend using fraction unless exact precision is necessary.

File: flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala (hunk 14 above)
Comment (ChengXiangLi) [inline flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala]: The validity of sample result is verified in after() method for each test. As the source data is very small, verify the fraction does not make much sense, so i didn't verify the fraction validity here, but it got verified in RandomSamplerTest in Sampler level by the way.

**Comment by ChengXiangLi** [comment]
Thanks for the review, @thvasilo , fixed all comments in latest commit.

File: flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala (hunk 14 above)
Comment (thvasilo) [inline flink-tests/src/test/scala/org/apache/flink/api/scala/operators/SampleITCase.scala]: Thanks, I missed that.

**Comment by ChengXiangLi** [comment]
Hi, @tillrohrmann , do you have sometime to continue to review this PR and help to push its progress?

File: flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+				return true;
+			}
+			
+			@Override
+			public T next() {
+				T result = currentElement;
+				if (currentCount == 0) {
+					currentElement = null;
+					return null;
+				}
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java]: Can we put the following statements in an else branch and remove the `T result = currentElement;`? I think this makes the control flow slightly more explicit.

Comment (ChengXiangLi) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java]: Thanks for the review, @tillrohrmann , I refactor this iterator, add else condition branch as you mentioned. The previous one lazily computer next element during hasNext(), which make this iterator only work while use invoke hasNext() at first, and then next(), fix this issue in latest commit as well.

Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/operators/util/PoissonSampler.java]: Great :-)

**Comment by tillrohrmann** [comment]
Sorry for the late review @ChengXiangLi. I finished it now and the PR looks really good. 

There is only one thing I would like to change before merging it. This is to move the `sample` and the `sampleWithSize` methods from the `DataSet` to the `DataSetUtils` class. This will effectively make the `sample` methods not part of the core API. The reason for this is that the `sampleWithSize` method breaks with one of Flink's guarantees, which is the robustness of the core API functions against `OutOfMemoryExceptions`. Let me elaborate on it for better understanding.

All of Flink internal operations work on serialized data which is stored in Flink's *managed* memory. The managed memory allows Flink to detect when to spill elements from memory to disk to avoid out of memory exceptions and, thus, to make the system robust. The managed memory is a pre-allocated area of memory which is administered by the `MemoryManager`. The `MemoryManager` allows you to allocate and deallocate `MemorySegments` in a c-style fashion. However, once a data item enters a UDF, the item has to be deserialized putting it on the remaining heap. This is not bad if your UDF does not accumulate these elements. However, the `sampleWithSize` method materializes up to `numSamples` elements on the heap. Depending on the number of samples and the data item size, this might be enough to eat up all remaining heap memory space and to crash the JVM.

I think that your current implementation will work for most use cases but in order to make it part of the core API, we also have to deal with the case where our sample cannot be materialized on the remaining heap of a running Flink program. In order to achieve this, I think it would be necessary to implement a native `topK` operator. With *native* I mean an operator which works on Flink's managed memory and, thus, can also deal with spilling records to disk. Having such a `topK` operator, we could reimplement the reservoir sampling algorithm the following way: For sampling without replacement we first assign weights in a map operation to each element. Then we call topK with respect to the weights and obtain the sample. For the sampling with replacement we could simply use a flat map operation to assign `numSamples` times a weight to each element. Then we again call `topK` with respect to the weight.

For the topK implementation, we would need something like a `PriorityQueue` which operates on managed memory (similar to the `CompactingHashTable` which is a hash table working on managed memory). Thus, we would have a priority queue which stores the priority values of each record and a pointer to the record which is kept in managed memory. Whenever an element is removed from the priority queue, we can also free the occupied managed memory. In case that we run out of managed memory, we have to spill some of the records to disk which are still in the race for the top k. As a first step, we can skip the spilling and just throw a proper exception (other than `OutOfMemoryException`) when we run out of memory. Afterwards, we can incrementally add the spilling functionality.

I know that you've already spent a lot of effort into writing the sampling operator and that this result might be a little bit demotivating. However, if we want to make it right and robust, then I think this is the way to go. Additionally we would add a proper topK operator to Flink's API which is missing big time :-) If you want to, then you could also take the lead here. The further discussion should then happen in a separate issue. I'm more than willing to assist you in implementing this operator. What do you think?

**Comment by ChengXiangLi** [comment]
Thanks for the detail explanation, @tillrohrmann . As the owner of this issue, make it work correctly and  efficiently is my top desire, so i would never say no to a reasonable and proper suggestion(although i did expect it happens earlier :-)). `OutOfMemoryError` is one of my concern during the implementation as well, unlimited usage of Java heap is unacceptable, so i limit the reservoir size(`PriorityQueue` size in implementation) same as `numSamples` parameter, which give the user full control of the memory used in java heap for `sample` operator. It may lead to `OutOfMemoryError` if the `numSamples` is too large, but other operators may lead to `OutOfMemoryError` as well if user cache too much objects inside UDF. Anyway, it's just my first thought, a fully managed memory usage is a better solution, and i would like to implement that.
As mentioned in the previous comment, is this the next step you prefer?
1. move the `sample` and the `sampleWithSize` methods from the `DataSet` to the `DataSetUtils` class, and merge this PR.
2. implement `topK` operator in a separate PR.
3. update `sampleWithSize` implementation, and move back to `DataSet`.

**Comment by tillrohrmann** [comment]
@ChengXiangLi, you're right, I should have noticed earlier and raise a flag. But your work is not in vain. I think it's some excellent piece of work and the `sample` method could also become part of the core API right away.

For the sake of completeness, let's do it once the `sampleWithSize` method works also robustly. I think your proposition for the next steps is a good way to continue with it. Once you've moved the `sample` and `sampleWithSize` methods to the `DataSetUtils` class, we close and merge this PR. In the meantime, I'll create a JIRA for the topK operator, where we can discuss the matter further.

**Comment by ChengXiangLi** [comment]
OK, @tillrohrmann , great to hear that. Besides, I've created [FLINK-2549](https://issues.apache.org/jira/browse/FLINK-2549) for topK operator.

**Comment by tillrohrmann** [comment]
Oh cool, then you were faster than me :-)

**Comment by ChengXiangLi** [comment]
Move sample/sampleWithSize operator to DataSetUtil and update unit test.

**Comment by tillrohrmann** [comment]
Looks great. Thanks a lot for your contribution @ChengXiangLi. Will merge it now.

**Comment by sachingoel0101** [comment]
Hey @ChengXiangLi , I just observed a failure on a test case: https://travis-ci.org/sachingoel0101/flink/jobs/76649177
Here is the relevant statement:
RandomSamplerTest.testPoissonSamplerFraction:116->verifySamplerFraction:249 expected fraction: 0.010000, result fraction: 0.009000

**Comment by thvasilo** [comment]
It's great to have this in, I'll try to update the cross-validation and SGD to use this.

**Comment by ChengXiangLi** [comment]
Hi, @sachingoel0101 , while sample with fraction, it's not easy to verify whether the DataSet is sampled with input fraction. In the test, i take 5 times sample, use the average size to computer the result fraction, and then compare the result fraction with input fraction, verify their difference is not more than 10% percent. The following case may happens as well, Sampler sample the DataSet with input fraction, but the sampled result size is too small or too large that beyond our verification condition, it happens, just with very little possibility, say less than 0.001 in this test. it should be ok if this failure happens very occasionally, please let me know if you found it's not.

**Comment by chiwanpark** [comment]
@ChengXiangLi I know that it is hard to verify random sampler implementation. But we need to fix this test failing because of difficulty of other pull requests verification. Some tests of other pull requests are failed by K-S test and sampling test with fraction. There is a [JIRA issue](https://issues.apache.org/jira/browse/FLINK-2564) covered this.

I'm testing with increased count of samples and source size. If I get a notable result, I'll post the result.

**Comment by ChengXiangLi** [comment]
Thanks, @chiwanpark , waiting for your result. 
Besides, we can reduce the fail-positive case to an acceptable possibility by expanding the verification boundary, while at the cost of making verification weak.

**Comment by sachingoel0101** [comment]
Hey @ChengXiangLi, I have another concern, regarding the seed for sampling. It doesn't seem to serve its purpose. I tried sampling with fraction three times with the same seed, however, every time I get different results. 
I've been stuck on this problem myself for quite some time. The fix is, instead of `seed + index`, you should just use `seed`. But then, the sample is not truly random. Far as I could figure out, splits of data don't arrive at the exactly same index subtask every time.

**Comment by tillrohrmann** [comment]
@thvasilo, the PR did not yet include the proper sampling behaviour within iterations. See FLINK-2396.

**Comment by tillrohrmann** [comment]
@sachingoel0101, you're right. The problem is that Flink does not give you a guarantee in which order the elements will arrive. But this problem won't be fixed by setting the seed for all sampling operators to the same value. There always might be an operator, e.g. rebalance, which will completely randomize your element order.

**Comment by sachingoel0101** [comment]
Yes. I was only wondering if we should at least ensure this when it is done right at the source though.

**Comment by tillrohrmann** [comment]
IMHO, this makes understanding the semantics of the sampling operator only more complicated because it behaves differently depending on the job graph structure. I would rather document this limitation more prominently.

**Pull request closed without merging**